jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java

```diff
@@ -497,44 +510,40 @@ void parseAndFillForContent()
 
     private int fillRequestBuffer()
     {
-        if (_retainableByteBuffer != null && _retainableByteBuffer.isRetained())
+        // We are not in a race here for the request buffer as we have not yet received a request,
+        // so there are not any possible legal threads calling #parseContent or #completed.
+
+        // We must fully consume the current buffer before filling again
+        if (!isRequestBufferEmpty())
+            return _requestBuffer.remaining();
+
+        // If the requestBuffer is retained, then we get a new one if the space available is limited
+        if (_requestBuffer != null && _requestBuffer.isRetained()  &&
+            BufferUtil.space(_requestBuffer.getByteBuffer()) < getHttpConfiguration().getOutputBufferSize() / 2)
         {
-            // TODO this is almost certainly wrong
             RetainableByteBuffer newBuffer = _bufferPool.acquire(getInputBufferSize(), isUseInputDirectByteBuffers());
             if (LOG.isDebugEnabled())
-                LOG.debug("replace buffer {} <- {} in {}", _retainableByteBuffer, newBuffer, this);
-            _retainableByteBuffer.release();
-            _retainableByteBuffer = newBuffer;
+                LOG.debug("replace buffer {} <- {} in {}", _requestBuffer, newBuffer, this);
+            _requestBuffer.release();
+            _requestBuffer = newBuffer;
         }
 
-        if (!isRequestBufferEmpty())
-            return _retainableByteBuffer.remaining();
-
         // Get a buffer
-        // We are not in a race here for the request buffer as we have not yet received a request,
-        // so there are not any possible legal threads calling #parseContent or #completed.
         ByteBuffer requestBuffer = getRequestBuffer();
 
-        // fill
         try
         {
             int filled = getEndPoint().fill(requestBuffer);
             if (filled == 0) // Do a retry on fill 0 (optimization for SSL connections)
                 filled = getEndPoint().fill(requestBuffer);
 
             if (LOG.isDebugEnabled())
-                LOG.debug("{} filled {} {}", this, filled, _retainableByteBuffer);
+                LOG.debug("filled {} {} {}", filled, this, _requestBuffer);
 
             if (filled > 0)
-            {
                 bytesIn.add(filled);
-            }
-            else
-            {
-                if (filled < 0)
-                    _parser.atEOF();
-                releaseRequestBuffer();
-            }
+            else if (filled < 0)
+                _parser.atEOF();
```

Comment:
Cannot remove buffer release from this method, because this method is called from `parseAndFillForContent()` (which does not release buffers), so reading the request content would never release the buffer.